u8timeseries/models/statistics.py

```diff
@@ -0,0 +1,136 @@
+from ..timeseries import TimeSeries
+import numpy as np
+from statsmodels.tsa.stattools import acf
+from scipy.stats import norm
+from statsmodels.tsa.seasonal import seasonal_decompose
+import matplotlib.pyplot as plt
+from typing import Tuple
+import math
+
+
+def check_seasonality(ts: 'TimeSeries', m: int = None, max_lag: int = 24, alpha: float = 5):
+    """
+    Returns whether the TimeSeries [ts] is seasonal with period [m] or not.
+
+    If [m] is None, the seasonality period is inferred from the Auto-correlation Function (ACF).
+
+    :param ts: The TimeSeries to check for seasonality.
+    :param m: The seasonality period to check.
+    :param max_lag: The maximal lag allowed in the ACF.
+    :param alpha: The desired confidence level (default 5%).
+    :return: A tuple (season, m), where s is a boolean indicating whether the TimeSeries has seasonality or not
+             and m is the seasonality period.
+    """
+    if m is not None:
+        if m < 0:
+            raise ValueError('m must be a positive integer.')
+
+    n_unique = np.unique(ts.values()).shape[0]
+
+    if n_unique > 1:  # Check for non-constant TimeSeries
+
+        r = acf(ts.values(), nlags=max_lag)  # In case user wants to check for seasonality higher than 24 steps.
+
+        grad = np.gradient(r)
+        signs_changes = np.diff(np.sign(grad))
+
+        if m is None:
+            # Tries to infer seasonality from AutoCorrelation Function if no value of m has been provided.
+            # We look for the first positive local maximum of the ACF by checking the sign changes in the gradient.
+
+            # Local maximum is indicated by signs_change == -2
+            if len(np.nonzero((signs_changes == -2))[0]) > 0:
+                m = np.nonzero((signs_changes == -2))[0][0] + 1
+            else:
+                return False, 0
+
+        # Check for local maximum when m is user defined
+        if np.nonzero((signs_changes == -2))[0][0] != m-1:
+            return False, m
+
+        # Remove r[0], the auto-correlation at order 0, that introduces bias.
+        r = r[1:]
+
+        stat = bartlett_formula(r, m - 1, len(ts))
```

Comment:
One that comes to mind, is that there are some technique/algo that require stationary time-series, so removing the trend can sometimes be helpful in making these time-series stationary.

Reply:
As far as I know Bartlett's formula is the most common one used for study of ACF, a few words could be said about that indeed.

Clearly it would be useful to also including de-trending as a tool in the library, @Digma's remark is very pertinent.

In terms of how effective it would be, in general, to first de-trend and then check seasonality, I believe it should not affect the result much since the underlying assumption with those two operations is that both trend and seasonality are deterministic. (Though @Bugdup, visually your example is very appealing!!)